### PR TITLE
feat: Add flag to control network requests breadcrumbs

### LINKED ISSUE #1336: Add an option to opt-out of HTTP breadcrumbs
We need a way for users to opt-out of having these breadcrumbs.

https://github.com/getsentry/sentry-cocoa/blob/468e7da4df2cb05bf29ec9d28ec9f2828be48e09/Sources/Sentry/SentryNetworkTracker.m#L144-L168

Documentation currently doesn't include `SentryNetworkTrackingIntegration`

https://docs.sentry.io/platforms/apple/configuration/integrations/default/

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Created a new flag to control whether network requests breadcrumb are created.
The default value is true. 
The new flag depends on swizzling being enabled.

## :bulb: Motivation and Context

closes #1336 

## :green_heart: How did you test it?

Unit tests

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x]  I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

File: Sources/Sentry/Public/SentryOptions.h [hunk 1]
```diff
@@ -64,6 +64,13 @@ NS_SWIFT_NAME(Options)
  */
 @property (nonatomic, assign) NSUInteger maxBreadcrumbs;
 
+/**
+ * Enable breadcrumbs to be created for each network request.
+ * Default value is YES.
+ * If enableSwizzling is disabled no breadcrumb will be created.
```
Comment (philipphofmann) [inline Sources/Sentry/Public/SentryOptions.h:70]: `m`:
```suggestion
 * When enabled, the SDK adds breadcrumbs for each network request. Default value is YES.
 * As this feature uses swizzling, disabling enableSwizzling also disables this feature.
```

File: Sources/Sentry/include/SentryNetworkTracker.h [hunk 2]
```diff
@@ -16,11 +16,14 @@ static NSString *const SENTRY_NETWORK_REQUEST_TRACKER_SPAN = @"SENTRY_NETWORK_RE
 
 - (nullable NSDictionary *)addTraceHeader:(nullable NSDictionary *)headers;
 
-- (void)enable;
+- (void)enableNetworkTracking;
+
+- (void)enableBreadcrumbs;
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentryNetworkTracker.h:21]: `l`: To be consistent with the naming above. Same applies, of course, for the property isBreadcrumbEnabled.
```suggestion
- (void)enableNetworkBreadcrumbs;
```

File: Sources/Sentry/SentryNetworkTracker.m [hunk 3]
```diff
@@ -175,6 +168,37 @@ - (void)urlSessionTask:(NSURLSessionTask *)sessionTask setState:(NSURLSessionTas
     [SentryLog logWithMessage:@"Finished HTTP span for sessionTask" andLevel:kSentryLevelDebug];
 }
 
+- (void)breadcrumbForSessionTask:(NSURLSessionTask *)sessionTask
```
Comment (philipphofmann) [inline Sources/Sentry/SentryNetworkTracker.m:171]: `l`: It totally makes sense to put this into an extra method. Maybe the following is a better method name.
```suggestion
- (void)addBreadcrumbForSessionTask:(NSURLSessionTask *)sessionTask
```

File: Sources/Sentry/SentryNetworkTrackingIntegration.m [hunk 4]
```diff
@@ -10,45 +10,49 @@ @implementation SentryNetworkTrackingIntegration
...
-    // SentryNetworkTracker (written on 29th of September 2021). Therefore we accept this tradeof
-    // for now.
+    if (!options.enableSwizzling) {
+        [SentryLog logWithMessage:
+                       @"Not going to enable NetworkTracking because enableSwizzling is disabled."
+                         andLevel:kSentryLevelDebug];
+        return;
+    }
+
+    BOOL needNetworkTracking = true;
```
Comment (philipphofmann) [inline Sources/Sentry/SentryNetworkTrackingIntegration.m:20]: `m`: `BOOL` is `YES`. Same applies for usages below. Maybe the following name suits better.
```suggestion
    BOOL shouldEnableNetworkTracking = YES;
```

File: Sources/Sentry/SentryNetworkTrackingIntegration.m [hunk 5]
```diff
@@ -10,45 +10,49 @@ @implementation SentryNetworkTrackingIntegration
 
 - (void)installWithOptions:(SentryOptions *)options
 {
-    // We are aware that the SDK only creates breadcrumbs for HTTP requests if performance is
```
Comment (philipphofmann) [inline Sources/Sentry/SentryNetworkTrackingIntegration.m:13]: Awesome, that we finally can remove this comment 🎉

File: Tests/SentryTests/SentryOptionsTest.m [hunk 6]
```diff
@@ -173,6 +173,16 @@ - (void)testMaxBreadcrumbs
...
 }
 
+- (void)testEnableNetworkBreadcrumbs
+{
+    NSNumber *enableNetworkBreadcrumbs = @NO;
+
+    SentryOptions *options =
+        [self getValidOptions:@{ @"enableNetworkBreadcrumbs" : enableNetworkBreadcrumbs }];
+
+    XCTAssertFalse(options.enableNetworkBreadcrumbs);
```
Comment (philipphofmann) [inline Tests/SentryTests/SentryOptionsTest.m:183]: `m`: That does the trick for you.
```suggestion
    [self testBooleanField:@"enableNetworkBreadcrumbs"];
```

File: Sources/Sentry/SentryNetworkTrackingIntegration.m [hunk 7]
```diff
@@ -10,45 +10,49 @@ @implementation SentryNetworkTrackingIntegration
...
+
+    if (options.enableNetworkBreadcrumbs) {
+        [SentryNetworkTracker.sharedInstance enableBreadcrumbs];
     }
 
-    [SentryNetworkTracker.sharedInstance enable];
-    [SentryNetworkTrackingIntegration swizzleNSURLSessionConfiguration];
-    [SentryNetworkTrackingIntegration swizzleURLSessionTask];
+    if (needNetworkTracking || options.enableNetworkBreadcrumbs) {
+        [SentryNetworkTrackingIntegration swizzleURLSessionTask];
```
Comment (philipphofmann) [inline Sources/Sentry/SentryNetworkTrackingIntegration.m:54]: `m`: I just realized that commenting out this line doesn't lead to a failing test in `SentryNetworkTrackerIntegrationTests`. Most likely this is my fault, but could you be so kind and add a test for this if possible, please?

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerIntegrationTests.swift [hunk 8]
```diff
@@ -66,6 +66,25 @@ class SentryNetworkTrackerIntegrationTests: XCTestCase {
         }
     }
     
+    func testBreadcrumbDisabled_WhenSwizzlingDisabled() {
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerIntegrationTests.swift:69]: Nice tests 👏

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift [hunk 9]
```diff
@@ -291,6 +292,36 @@ class SentryNetworkTrackerTests: XCTestCase {
         XCTAssertEqual(breadcrumb!.data!["response_body_size"] as! Int64, DATA_BYTES_RECEIVED)
     }
     
+    func testNoBreadcrumb_DisablingBreadcrumb() {
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:295]: `m`: I think we are missing a test case for `isNetworkTrackingEnabled = YES`, but `isBreadcrumbEnabled = FALSE`. I see we have one for both disabled, but not for the just mentioned case.

File: Sources/Sentry/SentryNetworkTrackingIntegration.m (hunk 7 above)
Comment (brustolin) [reply Sources/Sentry/SentryNetworkTrackingIntegration.m:54]: Sure.

File: Tests/SentryTests/SentryOptionsTest.m (hunk 6 above)
Comment (brustolin) [reply Tests/SentryTests/SentryOptionsTest.m:183]: nice

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift (hunk 9 above)
Comment (brustolin) [reply Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerTests.swift:295]: This `testNoBreadcrumb_DisablingBreadcrumb` test is exactly for that.
I first disable both, then enable just `isNetworkTrackingEnabled`

File: Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerIntegrationTests.swift (hunk 8 above)
Comment (brustolin) [reply Tests/SentryTests/Integrations/Performance/Network/SentryNetworkTrackerIntegrationTests.swift:69]: Thanks

File: Sources/Sentry/SentryNetworkTrackingIntegration.m (hunk 7 above)
Comment (brustolin) [reply Sources/Sentry/SentryNetworkTrackingIntegration.m:54]: Sure

**Review by philipphofmann — APPROVED** [review]
LGTM 🚀

**Pull request merged**